File: java/src/GzipDecompress.java
/* 
 * Simple DEFLATE decompressor
 * Copyright (c) Project Nayuki
 * 
 * https://www.nayuki.io/page/simple-deflate-decompressor
 * https://github.com/nayuki/Simple-DEFLATE-decompressor
 */

import java.io.BufferedInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInput;
import java.io.DataInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Date;
import java.util.zip.CRC32;
import java.util.zip.DataFormatException;


/**
 * Decompression application for the gzip file format.
 * <p>Usage: java GzipDecompress InputFile.gz OutputFile</p>
 * <p>This decompresses a single gzip input file into a single output file. The program also prints
 * some information to standard output, and error messages if the file is invalid/corrupt.</p>
 */
public final class GzipDecompress {
	
	public static void main(String[] args) {
		String msg = submain(args);
		if (msg != null) {
			System.err.println(msg);
			System.exit(1);
		}
	}
	
	
	// Returns null if successful, otherwise returns an error message string.
	private static String submain(String[] args) {
		// Handle command line arguments
		if (args.length != 2)
			return "Usage: java GzipDecompress InputFile.gz OutputFile";
		File inFile = new File(args[0]);
		if (!inFile.exists())
			return "Input file does not exist: " + inFile;
		if (inFile.isDirectory())
			return "Input file is a directory: " + inFile;
		Path outFile = Paths.get(args[1]);
		
		try {
			byte[] decomp;
			int crc, size;
			// Start reading
			try (DataInputStream in = new DataInputStream(new BufferedInputStream(new FileInputStream(inFile), 16 * 1024))) {
				// Header
				int flags;
				{
					if (in.readUnsignedShort() != 0x1F8B)
						return "Invalid GZIP magic number";
					int compMeth = in.readUnsignedByte();
					if (compMeth != 8)
						return "Unsupported compression method: " + compMeth;
					flags = in.readUnsignedByte();
					
					// Reserved flags
					if ((flags & 0xE0) != 0)
						return "Reserved flags are set";
					
					// Modification time
					int mtime = readLittleEndianInt32(in);
					if (mtime != 0)
						System.out.println("Last modified: " + new Date(mtime * 1000L));
					else
						System.out.println("Last modified: N/A");
					
					// Extra flags
					int extraFlags = in.readUnsignedByte();
					switch (extraFlags) {
						case 2:   System.out.println("Extra flags: Maximum compression");  break;
						case 4:   System.out.println("Extra flags: Fastest compression");  break;
						default:  System.out.println("Extra flags: Unknown (" + extraFlags + ")");  break;
					}
					
					// Operating system
					String os;
					switch (in.readUnsignedByte()) {
						case   0:  os = "FAT";             break;
						case   1:  os = "Amiga";           break;
						case   2:  os = "VMS";             break;
						case   3:  os = "Unix";            break;
						case   4:  os = "VM/CMS";          break;
						case   5:  os = "Atari TOS";       break;
						case   6:  os = "HPFS";            break;
						case   7:  os = "Macintosh";       break;
						case   8:  os = "Z-System";        break;
						case   9:  os = "CP/M";            break;
						case  10:  os = "TOPS-20";         break;
						case  11:  os = "NTFS";            break;
						case  12:  os = "QDOS";            break;
						case  13:  os = "Acorn RISCOS";    break;
						case 255:  os = "Unknown";         break;
						default :  os = "Really unknown";  break;
					}
					System.out.println("Operating system: " + os);
				}
				
				// Handle assorted flags
				if ((flags & 0x01) != 0)
					System.out.println("Flag: Text");
				if ((flags & 0x04) != 0) {
					System.out.println("Flag: Extra");
					int len = readLittleEndianUint16(in);
					in.readFully(new byte[len]);  // Skip extra data
				}
				if ((flags & 0x08) != 0)
					System.out.println("File name: " + readNullTerminatedString(in));
				if ((flags & 0x02) != 0)
					System.out.printf("Header CRC-16: %04X%n", readLittleEndianUint16(in));
				if ((flags & 0x10) != 0)
					System.out.println("Comment: " + readNullTerminatedString(in));
				
				// Decompress
				try {
					decomp = Decompressor.decompress(new ByteBitInputStream(in));
				} catch (DataFormatException e) {
					return "Invalid or corrupt compressed data: " + e.getMessage();
				}
				
				// Footer
				crc  = readLittleEndianInt32(in);
				size = readLittleEndianInt32(in);
			}
			
			// Check decompressed data's length and CRC
			if (size != decomp.length)
				return String.format("Size mismatch: expected=%d, actual=%d", size, decomp.length);
			if (crc != getCrc32(decomp))
				return String.format("CRC-32 mismatch: expected=%08X, actual=%08X", crc, getCrc32(decomp));
			
			// Write decompressed data to output file
			Files.write(outFile, decomp);
			
		} catch (IOException e) {
			return "I/O exception: " + e.getMessage();
		}
		return null;  // Success, no error message
	}
	
	
	/*---- Helper methods ----*/
	
	private static String readNullTerminatedString(DataInput in) throws IOException {
		ByteArrayOutputStream bout = new ByteArrayOutputStream();
		while (true) {
			byte b = in.readByte();
			if (b == 0)
				break;
			bout.write(b);
		}
		return new String(bout.toByteArray(), StandardCharsets.UTF_8);
	}
	
	
	private static int getCrc32(byte[] data) {
		CRC32 crc = new CRC32();
		crc.update(data);
		return (int)crc.getValue();
	}
	
	
	private static int readLittleEndianUint16(DataInput in) throws IOException {
		return Integer.reverseBytes(in.readUnsignedShort()) >>> 16;
	}
	
	
	private static int readLittleEndianInt32(DataInput in) throws IOException {
		return Integer.reverseBytes(in.readInt());
	}
	
}

File: cpp/GzipDecompress.cpp
/* 
 * Simple DEFLATE decompressor (C++)
 * 
 * Copyright (c) Project Nayuki
 * MIT License. See readme file.
 * https://www.nayuki.io/page/simple-deflate-decompressor
 */

#include <bitset>
#include <cstdlib>
#include <exception>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <istream>
#include <string>
#include <vector>
#include "DeflateDecompress.hpp"

using std::uint8_t;
using std::uint16_t;
using std::uint32_t;
using std::string;


class DataInput final {
	
	private: std::istream &input;
	
	
	public: DataInput(std::istream &in) :
		input(in) {}
	
	
	public: uint8_t readUint8() {
		int b = input.get();
		if (b == std::char_traits<char>::eof())
			throw std::runtime_error("Unexpected end of stream");
		return static_cast<uint8_t>(b);
	}
	
	
	public: uint16_t readLittleEndianUint16() {
		uint16_t result = 0;
		for (int i = 0; i < 2; i++)
			result |= static_cast<uint16_t>(readUint8()) << (i * 8);
		return result;
	}
	
	
	public: uint32_t readLittleEndianUint32() {
		uint32_t result = 0;
		for (int i = 0; i < 4; i++)
			result |= static_cast<uint32_t>(readUint8()) << (i * 8);
		return result;
	}
	
	
	public: string readNullTerminatedString() {
		string result;
		while (true) {
			int b = input.get();
			if (b == std::char_traits<char>::eof())
				throw std::runtime_error("Unexpected end of stream");
			else if (b == '\0')
				break;
			else
				result.push_back(static_cast<char>(b));
		}
		return result;
	}
	
};


static uint32_t getCrc32(const std::vector<uint8_t> &data) {
	uint32_t crc = ~UINT32_C(0);
	for (uint8_t b : data) {
		crc ^= b;
		for (int i = 0; i < 8; i++)
			crc = (crc >> 1) ^ ((crc & 1) * UINT32_C(0xEDB88320));
	}
	return ~crc;
}


static string toHex(uint32_t val, int digits) {
	std::ostringstream s;
	s << std::hex << std::setw(digits) << std::setfill('0') << val;
	return s.str();
}


static string submain(int argc, char *argv[]) {
	// Handle command line arguments
	if (argc != 3)
		return string("Usage: ") + argv[0] + " GzipDecompress InputFile.gz OutputFile";
	const char *inFile = argv[1];
	if (!std::filesystem::exists(inFile))
		return string("Input file does not exist: ") + inFile;
	if (std::filesystem::is_directory(inFile))
		return string("Input file is a directory: ") + inFile;
	const char *outFile = argv[2];
	
	try {
		std::vector<uint8_t> decomp;
		uint32_t crc, size;
		
		// Start reading
		{
			std::ifstream in0(inFile);
			DataInput in1(in0);
			
			// Header
			std::bitset<8> flags;
			{
				if (in1.readLittleEndianUint16() != 0x8B1F)
					return "Invalid GZIP magic number";
				int compMeth = in1.readUint8();
				if (compMeth != 8)
					return string("Unsupported compression method: ") + std::to_string(compMeth);
				flags = in1.readUint8();
				
				// Reserved flags
				if (flags[5] || flags[6] || flags[7])
					return "Reserved flags are set";
				
				// Modification time
				uint32_t mtime = in1.readLittleEndianUint32();
				if (mtime != 0)
					std::cout << "Last modified: " << mtime << " (Unix time)" << std::endl;
				else
					std::cout << "Last modified: N/A";
				
				// Extra flags
				std::cout << "Extra flags: ";
				int extraFlags = in1.readUint8();
				switch (extraFlags) {
					case 2:   std::cout << "Maximum compression";  break;
					case 4:   std::cout << "Fastest compression";  break;
					default:  std::cout << "Unknown (" << extraFlags << ")";  break;
				}
				std::cout << std::endl;
				
				// Operating system
				int operatingSystem = in1.readUint8();
				string os;
				switch (operatingSystem) {
					case   0:  os = "FAT";           break;
					case   1:  os = "Amiga";         break;
					case   2:  os = "VMS";           break;
					case   3:  os = "Unix";          break;
					case   4:  os = "VM/CMS";        break;
					case   5:  os = "Atari TOS";     break;
					case   6:  os = "HPFS";          break;
					case   7:  os = "Macintosh";     break;
					case   8:  os = "Z-System";      break;
					case   9:  os = "CP/M";          break;
					case  10:  os = "TOPS-20";       break;
					case  11:  os = "NTFS";          break;
					case  12:  os = "QDOS";          break;
					case  13:  os = "Acorn RISCOS";  break;
					case 255:  os = "Unknown";       break;
					default :  os = string("Really unknown (") + std::to_string(operatingSystem) + ")";  break;
				}
				std::cout << "Operating system: " << os << std::endl;
			}
			
			// Handle assorted flags
			if (flags[0])
				std::cout << "Flag: Text" << std::endl;
			if (flags[2]) {
				std::cout << "Flag: Extra" << std::endl;
				long len = in1.readLittleEndianUint16();
				for (long i = 0; i < len; i++)  // Skip extra data
					in1.readUint8();
			}
			if (flags[3])
				std::cout << "File name: " + in1.readNullTerminatedString() << std::endl;
			if (flags[1])
				std::cout << "Header CRC-16: " << toHex(in1.readLittleEndianUint16(), 4) << std::endl;
			if (flags[4])
				std::cout << "Comment: " + in1.readNullTerminatedString() << std::endl;
			
			// Decompress
			try {
				BitInputStream in2(in0);
				decomp = Decompressor::decompress(in2);
			} catch (std::exception &e) {
				return string("Invalid or corrupt compressed data: ") + e.what();
			}
			
			// Footer
			crc  = in1.readLittleEndianUint32();
			size = in1.readLittleEndianUint32();
		}
		
		// Check decompressed data's length and CRC
		if (size != static_cast<uint32_t>(decomp.size()))
			return string("Size mismatch: expected=") + std::to_string(size) + ", actual=" + std::to_string(decomp.size());
		if (crc != getCrc32(decomp))
			return string("CRC-32 mismatch: expected=") + toHex(crc, 8) + ", actual=" + toHex(getCrc32(decomp), 8);
		
		// Write decompressed data to output file
		std::ofstream out(outFile);
		for (uint8_t b : decomp)
			out.put(b);
		
		// Success, no error message
		return "";
		
	} catch (std::exception &e) {
		return string("I/O exception: ") + e.what();
	}
}


/* 
 * Decompression application for the gzip file format.
 * Usage: GzipDecompress InputFile.gz OutputFile
 * This decompresses a single gzip input file into a single output file. The program also prints
 * some information to standard output, and error messages if the file is invalid/corrupt.
 */
int main(int argc, char *argv[]) {
	string msg = submain(argc, argv);
	if (msg.length() == 0)
		return EXIT_SUCCESS;
	else {
		std::cerr << msg << std::endl;
		return EXIT_FAILURE;
	}
}
